app/src/main/java/org/mozilla/tv/firefox/ScreenController.kt

```diff
@@ -53,53 +79,113 @@ class ScreenController {
-        // method is called from Settings.
-        //
-        // However, from a user perspective, the behavior is correct (e.g. back stack functions
-        // correctly with multiple sessions).
-        val browserFragment = fragmentManager.findFragmentByTag(WebRenderFragment.FRAGMENT_TAG) as? WebRenderFragment
-        if (browserFragment != null && browserFragment.isVisible) {
-            // We can't call loadUrl on the Fragment until the view hierarchy is inflated so we check
-            // for visibility in addition to existence.
-            browserFragment.loadUrl(url)
+    // TODO: handle about:home and close overlay by button press separately
+    fun showNavigationOverlay(fragmentManager: FragmentManager?, toShow: Boolean) {
+        fragmentManager ?: return
+        var transaction = fragmentManager.beginTransaction()
+        val overlayFragment = fragmentManager.navigationOverlayFragment()
+        val renderFragment = fragmentManager.webRenderFragment()
+        if (toShow) {
+            stateMachine.overlayOpened()
+            transaction = transaction.show(overlayFragment)
+                .hide(renderFragment) // TODO note that this will need to be changed in order to display WebRenderFragment under a split overlay
         } else {
-            val session = Session(url, source = source)
-            context.webRenderComponents.sessionManager.add(session, selected = true)
+            stateMachine.overlayClosed()
+            transaction = transaction.hide(overlayFragment)
+                .show(renderFragment)
         }
+        transaction.commit()
     }
 
-    fun showPocketScreen(fragmentManager: FragmentManager) {
-        fragmentManager.beginTransaction()
-                .replace(R.id.container, PocketVideoFragment(), PocketVideoFragment.FRAGMENT_TAG)
-                .addToBackStack(null)
-                .commit()
+    fun handleBack(fragmentManager: FragmentManager): Boolean {
+        if (stateMachine.currentActiveScreen == ScreenControllerStateMachine.ActiveScreen.WEB_RENDER) {
+            val webRenderFragment = fragmentManager.webRenderFragment()
+            if (webRenderFragment.onBackPressed()) return true
+        }
+        return handleTransition(fragmentManager, stateMachine.backPress())
+    }
+
+    fun handleMenu(fragmentManager: FragmentManager) {
+        handleTransition(fragmentManager, stateMachine.menuPress())
+    }
+
+    private fun handleTransition(fragmentManager: FragmentManager, transition: ScreenControllerStateMachine.Transition): Boolean {
+        when (transition) {
+            ScreenControllerStateMachine.Transition.ADD_OVERLAY -> showNavigationOverlay(fragmentManager, true)
+            ScreenControllerStateMachine.Transition.REMOVE_OVERLAY -> showNavigationOverlay(fragmentManager, false)
+            ScreenControllerStateMachine.Transition.ADD_POCKET -> {
+                fragmentManager.beginTransaction()
```

Comment:
True - I also like that they are all in one place, because they're tied to the transition. Updated the method name to reflect that this is handling the fragment transition.